### PR TITLE
Add granular error messages to Dict.validate (bokeh#13295)

### LINKED ISSUE #13295: Non-specific error when the DataFrame Used to construct a ColumnDataSource contains non-string keys
### Software versions

Python version      :  3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:20:46)
IPython version     :  8.14.0
Tornado version     :  6.3.2
Bokeh version       :  3.2.1
BokehJS static path :  /opt/tljh/user/lib/python3.9/site-packages/bokeh/server/static
node.js version     :  (not installed)
npm version         :  (not installed)
Operating system    :  Linux-6.1.0-10-amd64-x86_64-with-glibc2.36

Consider the following example

```python
import pandas
from bokeh.models import ColumnDataSource
df = pandas.DataFrame({'a':[1,2,3], 1:[1,2,3]})
source = ColumnDataSource(df)
```

It will fail with an error `ValueError: expected an element of ColumnData(String, Seq(Any)), got {'index': array([0, 1, 2]), 1: array([1, 2, 3]), 'a': array([1, 2, 3])}` which tells the user nothing about the source of the problem. This problem is particularly obfuscated when creating large dataframes from dynamic sources.

What's worse is that the user will be pointed to a place that's [not the origin of the problem](https://github.com/bokeh/bokeh/blob/branch-3.3/src/bokeh/core/property/bases.py#L364) when the error actually happens [elsewhere](https://github.com/bokeh/bokeh/blob/branch-3.3/src/bokeh/core/property/container.py#L205C14-L205C14).

Even if the error correctly traced the issue to `bokeh.core.property.container.Dict` the method `validate` conflates the issue of `isinstance` with all invalid key errors and all invalid value errors.

### Browser name and version

_No response_

### Jupyter notebook / Jupyter Lab version

_No response_

### Expected behavior

A descriptive error should be raised explaining which key validation fails and why

### Observed behavior

The following message is raised:

ValueError: expected an element of ColumnData(String, Seq(Any)), got {'index': array([0, 1, 2]), 1: array([1, 2, 3]), 'a': array([1, 2, 3])}

### Example code

```Python
import pandas
from bokeh.models import ColumnDataSource
df = pandas.DataFrame({'a':[1,2,3], 1:[1,2,3]})
source = ColumnDataSource(df)
```

### Stack traceback or browser console output

_No response_

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by marceloslacerda**
- [X] issues: fixes #13295)
- [X] tests added / passed

### CODE REVIEW

File: src/bokeh/core/property/container.py [hunk 1]
```diff
@@ -202,11 +202,14 @@ def validate(self, value: Any, detail: bool = True) -> None:
...
-        if isinstance(value, dict) and all(key_is_valid(key) and value_is_valid(val) for key, val in value.items()):
-            return
-
-        msg = "" if not detail else f"expected an element of {self}, got {value!r}"
-        raise ValueError(msg)
+        expected = f"expected a dict of type {self}"
+        if not isinstance(value, dict):
+            raise ValueError(f"{expected}, got a value of type {type(value)}" if detail else "")
+        if any(bad_keys := [str(k) for k in value if not key_is_valid(k)]):
+            raise ValueError(f"{expected}, got a dict with invalid keys: {','.join(bad_keys)}" if detail else "")
```
Comment (bryevdv) [inline src/bokeh/core/property/container.py]:
```suggestion
        expected = f"expected a dict of type {self}"
        
        if not isinstance(value, dict):
            raise ValueError(f"{expected}, got a value of type {type(value)}" if detail else "")
            
        if any(bad_keys := [str(k) for k in value if not key_is_valid(k)]):
            raise ValueError(f"{expected}, got a dict with invalid keys: {','.join(bad_keys)}" if detail else "")
```
@marceloslacerda this is what I meant by vertical space: visual separation of blocks of code

File: tests/unit/bokeh/core/property/test_container.py [hunk 2]
```diff
@@ -144,6 +144,29 @@ def test_init(self) -> None:
         with pytest.raises(TypeError):
             bcpc.Dict()
 
+    def test_validate_invalid_key(self) -> None:
+        d = bcpc.Dict(String, String)
+        with pytest.raises(ValueError) as err:
+            d.validate({"Foo": "Bar", 1: "Baz"})
+            assert("invalid keys: 1" in err.value.args[0])
```
Comment (bryevdv) [inline tests/unit/bokeh/core/property/test_container.py]: No need for parens with asserts (here and below):
```suggestion
            assert "invalid keys: 1" in err.value.args[0]
```
It might be worth seeing if there is a simpler express, e.g. would ` "invalid keys: 1" in str(err.value)` work? (not sure, just a guess)

File: tests/unit/bokeh/core/property/test_container.py [hunk 3]
```diff
@@ -144,6 +144,29 @@ def test_init(self) -> None:
...
+    def test_validate_invalid_arg(self) -> None:
+        d = bcpc.Dict(String, String)
+        with pytest.raises(ValueError) as err:
+            d.validate([])
+            assert("got a value of type list" in err.value.args[0])
+
+    def test_validate_valid(self) -> None:
+        d = bcpc.Dict(String, String)
+        d.validate({"Foo": "Bar"})
+        assert True
```
Comment (bryevdv) [inline tests/unit/bokeh/core/property/test_container.py]: I like having positive tests, but the actual check here is that no exception is raised. I think you can use:
```python
with pyest.does_not_raise():
```
to test for that explicitly.

File: tests/unit/bokeh/core/property/test_validation__property.py [hunk 4]
```diff
@@ -175,7 +175,25 @@ def test_Dict(self) -> None:
...
+        d = Dict(String, String)
+        with pytest.raises(ValueError) as err:
+            d.validate({"Foo": "Bar", "Baz": 1})
+            assert("values for keys: Baz" in err.value.args[0])
+
+    def test_Dict_Valid(self) -> None:
+        d = Dict(String, String)
+        d.validate({"Foo": "Bar"})
+        assert True
+
```
Comment (bryevdv) [inline tests/unit/bokeh/core/property/test_validation__property.py:221]: same comments in this file as in `test_container.py`

File: tests/unit/bokeh/core/property/test_container.py (hunk 3 above)
Comment (marceloslacerda) [reply tests/unit/bokeh/core/property/test_container.py]: That function doesn't seem to exist

File: tests/unit/bokeh/core/property/test_container.py (hunk 2 above)
Comment (marceloslacerda) [reply tests/unit/bokeh/core/property/test_container.py]: `str(err) == "<ExceptionInfo ValueError('expected a dict of type Dict(String, String), got a dict with invalid keys: 1') tblen=2>"`

`str(err.value) == "expected a dict of type Dict(String, String), got a dict with invalid keys: 1"`

I'm going with the second option unless if you want otherwise

File: tests/unit/bokeh/core/property/test_container.py (hunk 3 above)
Comment (bryevdv) [reply tests/unit/bokeh/core/property/test_container.py]: sorry I guess the first refs I found skipped some details, should have just gone straight to the docs:

https://docs.pytest.org/en/7.1.x/example/parametrize.html#parametrizing-conditional-raising

TLDR, it's actually: `from contextlib import nullcontext as does_not_raise`

File: tests/unit/bokeh/core/property/test_container.py (hunk 2 above)
Comment (bryevdv) [reply tests/unit/bokeh/core/property/test_container.py]: 👍 sounds good!

File: tests/unit/bokeh/core/property/test_validation__property.py [hunk 5]
```diff
@@ -175,7 +175,27 @@ def test_Dict(self) -> None:
...
+        d = Dict(String, String)
+        with pytest.raises(ValueError) as err:
+            d.validate({"Foo": "Bar", 1: "Baz"})
+        assert "invalid keys: 1" in str(err.value)
+
+    def test_Dict_Invalid_Value(self) -> None:
+        d = Dict(String, String)
+        with pytest.raises(ValueError) as err:
+            d.validate({"Foo": "Bar", "Baz": 1})
+        assert "values for keys: Baz" in str(err.value)
```
Comment (bryevdv) [inline tests/unit/bokeh/core/property/test_validation__property.py]: If you are up for one more small push, I'd advise adding the following cases:

* multiple invalid keys
* multiple invalid values
* invalid keys and values at the same time

File: tests/unit/bokeh/core/property/test_validation__property.py [hunk 6]
```diff
@@ -175,7 +175,27 @@ def test_Dict(self) -> None:
...
+
+    def test_Dict_Valid(self) -> None:
+        d = Dict(String, String)
+        try:
+            d.validate({"Foo": "Bar"})
```
Comment (bryevdv) [inline tests/unit/bokeh/core/property/test_validation__property.py]: We could also consider really expanding the validation of correct cases via pytests `parametrize` decorator, something like (untested):
```suggestion
    KEYS = ((String, "Foo"), ("Int", 10))
    VALS = ((String, "Bar"), ("Int", 20), (Float, 1.23), (List[Float], [1.23, 3.45]))

    @pytest.mark.parametrize("key_type,key", KEYS)
    @pytest.mark.parametrize("val_type,val", VALS)
    def test_Dict_Valid(self, key_type, key, val_type, val) -> None:
        d = Dict(key_type, val_type)
        try:
            d.validate({key: val})
```

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Hi @marceloslacerda this looks good to me. I did have some final suggestions for some additional tests that I think would be nice-to-haves. Let me know what you think, we can merge this as-is if you don't have bandwidth for more work on this PR.

**Comment by marceloslacerda** [comment]
Sure bryan. I’ll be out of town this weekend so, if you can wait, I’ll be
back at working on this next week.

Em qui., 10 de ago. de 2023 às 18:04, Bryan Van de Ven <
***@***.***> escreveu:

**Comment by bryevdv** [comment]
👍 thanks @marceloslacerda !

File: tests/unit/bokeh/core/property/test_validation__property.py (hunk 5 above)
Comment (marceloslacerda) [reply tests/unit/bokeh/core/property/test_validation__property.py]: The multiple invalid keys and values at the same time is equal to the multiple invalid keys. Do you want me to change the `Dict.validate` code to account for that case?

File: tests/unit/bokeh/core/property/test_validation__property.py (hunk 6 above)
Comment (marceloslacerda) [reply tests/unit/bokeh/core/property/test_validation__property.py]: Sounds interesting, I'll give it a shot

**Comment by marceloslacerda** [comment]
@bryevdv check the latest errors

` E       ValueError: expected a Property as type parameter, got bokeh.core.property.container.List[bokeh.core.property.primitive.Float]`

How the heck is that happening?

File: tests/unit/bokeh/core/property/test_validation__property.py (hunk 5 above)
Comment (bryevdv) [reply tests/unit/bokeh/core/property/test_validation__property.py]: I do think the best UX for users would provide all the errors in one messages in a reasonably compact way, i.e. to avoid running, fixing some errors, running again, fixing more errors, etc.

**Comment by bryevdv** [comment]
@marceloslacerda I believe it's actually `List(Float)` not `List[Float]` (my mistake in the untested suggestion snippet)

**Comment by marceloslacerda** [comment]
I implemented the changes you suggested, Dict.validate is looking quite verbose now :-D

File: src/bokeh/core/property/container.py [hunk 7]
```diff
@@ -202,11 +202,25 @@ def validate(self, value: Any, detail: bool = True) -> None:
...
+        expected = f"expected a dict of type {self}"
+
+        if not isinstance(value, dict):
+            raise ValueError(f"{expected}, got a value of type {type(value)}" if detail else "")
+
+        bad_keys = [str(k) for k in value if not key_is_valid(k)]
+        bad_value_keys = [str(k) for (k, v) in value.items() if not value_is_valid(v)]
+        exception_header = f"{expected}, got a dict with"
+        bad_keys_str = f"invalid keys: {', '.join(bad_keys)}"
+        bad_key_values_str = f"invalid values for keys: {', '.join(bad_value_keys)}"
```
Comment (bryevdv) [inline src/bokeh/core/property/container.py]: My only comment is that this naming seems inconsistent, I'd expect `bad_value_keys -> bad_value_keys_str` based on `bad_keys -> bad_keys_str`

**Review by bryevdv — APPROVED** [review]
Looks great @marceloslacerda I left one optional nit comment but I will plan to merge tonight in whatever state things are!

File: src/bokeh/core/property/container.py (hunk 7 above)
Comment (marceloslacerda) [reply src/bokeh/core/property/container.py]: Shoot, you are right. If you want to wait I can patch that when I get home.

Comment (bryevdv) [reply src/bokeh/core/property/container.py]: sure I'll merge when you give you 👍

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Test failure is spurious, merging now. Thanks @marceloslacerda !
